pennylane/data/utils.py

```diff
@@ -203,122 +201,119 @@ def _check_data_exist(data_type, description, directory_path):
+def _s3_download(data_type, folders, attributes, dest_folder):
+    """Download a file for each attribute from each folder to the specified destination."""
+    for folder in folders:
+        s3_folder = os.path.join(S3_URL, data_type, folder)
+        local_folder = os.path.join(dest_folder, data_type, folder)
+        if not os.path.exists(local_folder):
+            os.makedirs(local_folder)
+
+        file_prefix = folder.replace("/", "_")
+        # TODO: if len(attributes) > 1, merge contents to single file like _partial.dat
+        for attr in attributes:
+            fname = f"{file_prefix}_{attr}.dat"
+            response = requests.get(f"{s3_folder}/{fname}", timeout=5.0)
+            if not response.ok:
+                response.raise_for_status()
+            with open(f"{local_folder}/{fname}", "wb") as f:
+                f.write(response.content)
+
+
+def _generate_folder_list(node, folders):
+    """Recursive helper method for _generate_folders"""
+    next_folders = folders[1:]
+    folders = set(node) if folders[0] == ["full"] else set(folders[0]).intersection(set(node))
+    if not next_folders:
+        return folders
+    return [
+        os.path.join(folder, child)
+        for folder in folders
+        for child in _generate_folder_list(node[folder], next_folders)
+    ]
+
+
+def _generate_folders(data_type, folders):
+    """Recursively generate and return a tree of all folder names matching a pattern.
+
+    If the __s3_filemap is stale, this method will refresh it.
+
+    Args:
+        data_type (str): The data type for which folders are generated
+        folders: (list[list[str]]): The ordered list of folder names requested.
+            The value ``["full"]`` will expand to all possible folders at that depth
+
+    Returns:
+        list[str]: The paths of files that should be fetched from S3
+    """
+
+    global _s3_filemap
+    if data_type not in _s3_filemap:
+        _s3_filemap = list_datasets()
+    return _generate_folder_list(_s3_filemap[data_type], folders)
+
+
+def load(data_type, attributes=None, lazy=False, folder_path="", force=True, **params):
     r"""Downloads the data if it is not already present in the directory and return it to user as a Datset object
 
     Args:
         data_type (str):  A string representing the type of the data required - qchem or qspin
-        description (dict): A dictionary with parameters for the required type of data
         attributes (list): An optional list to specify individual data element that are required
         folder_path (str): Path to the root folder where download takes place. By default dataset folder will be created in the working directory.
         force (Bool): Bool representing whether data has to be downloaded even if it is still present
+        params (kwargs): Keyword arguments exactly matching the parameters required for the data type. Note that these are not optional
```

Comment:
So a user will enter, something like: load("qspin", sysname='Heisenberg',layout='1x16', periodicity='periodic',lattice='chain')? I like this!